clippy_lints/src/await_holding_lock.rs

```diff
@@ -0,0 +1,100 @@
+use crate::utils::span_lint_and_note;
+use if_chain::if_chain;
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{Body, FnDecl, HirId, IsAsync};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{Span, Symbol};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for calls to await while holding a MutexGuard.
+    ///
+    /// **Why is this bad?** This is almost certainly an error which can result
+    /// in a deadlock because the reactor will invoke code not visible to the
+    /// currently visible scope.
+    ///
+    /// **Known problems:** Detects only specifically named guard types:
+    /// MutexGuard, RwLockReadGuard, and RwLockWriteGuard.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// use std::sync::Mutex;
+    ///
+    /// async fn foo(x: &Mutex<u32>) {
+    ///   let guard = x.lock().unwrap();
+    ///   *guard += 1;
+    ///   bar.await;
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::sync::Mutex;
+    ///
+    /// async fn foo(x: &Mutex<u32>) {
+    ///   {
+    ///     let guard = x.lock().unwrap();
+    ///     *guard += 1;
+    ///   }
+    ///   bar.await;
+    /// }
+    /// ```
+    pub AWAIT_HOLDING_LOCK,
+    pedantic,
+    "Inside an async function, holding a MutexGuard while calling await"
+}
+
+const MUTEX_GUARD_TYPES: [&str; 3] = ["MutexGuard", "RwLockReadGuard", "RwLockWriteGuard"];
```

Comment:
These should be full paths in util/paths.rs

Reply:
Then we can just use match_def_path

Reply:
I went this route to capture:

```
std::sync::mutex::MutexGuard
futures_locks::MutexGuard
parking_lot::MutexGuard
```

with one check. Similarly for the others.

I will change to using full paths as that certainly reduces false positives, but at the cost of false negatives for custom guard types.

Reply:
Do we want this in the first place for custom guard types from async executors (?)

Reply:
I initially thought yes you wanted this for async-aware locks too, but that seems to be an incorrect assumption on my part. We can leave it to std::sync and parking_lot:: as there does not appear to be any disagreement on that being wrong.